src/containers/containers.c

```diff
@@ -289,6 +293,340 @@ extern inline container_t *container_andnot(
+        case ARRAY_CONTAINER_TYPE: {
+            const array_container_t *ac = const_CAST_array(c);
+            it->in_container_index++;
+            it->has_value = it->in_container_index < ac->cardinality;
+            if (it->has_value) {
+                it->value = ac->array[it->in_container_index];
+            }
+            return it->has_value;
+        }
+        case RUN_CONTAINER_TYPE: {
+            if (it->value == UINT16_MAX) {  // Avoid overflow to zero
+                return (it->has_value = false);
+            }
+
+            const run_container_t *rc = const_CAST_run(c);
+            uint32_t limit =
+                rc->runs[it->run_index].value + rc->runs[it->run_index].length;
+            it->value++;
+            if (it->value <= limit) {
+                return (it->has_value = true);
+            }
+
+            it->run_index++;
+            it->has_value = it->run_index < rc->n_runs;
+            if (it->has_value) {
+                it->value = rc->runs[it->run_index].value;
+            }
+            return it->has_value;
+        }
+        default:
+            assert(false);
+            roaring_unreachable;
+            return false;
+    }
+}
+
+bool container_iterator_prev(const container_t *c, uint8_t typecode,
+                             roaring_container_iterator_t *it) {
+    switch (typecode) {
+        case BITSET_CONTAINER_TYPE: {
+            if (--it->in_container_index < 0) {
+                return (it->has_value = false);
+            }
+
+            const bitset_container_t *bc = const_CAST_bitset(c);
+            int32_t wordindex = it->in_container_index / 64;
+            uint64_t word =
+                bc->words[wordindex] &
+                (UINT64_MAX >> (63 - (it->in_container_index % 64)));
+
+            while (word == 0 && --wordindex >= 0) {
+                word = bc->words[wordindex];
+            }
+            if (word == 0) {
+                return (it->has_value = false);
+            }
+
+            it->in_container_index =
+                (wordindex * 64) + (63 - roaring_leading_zeroes(word));
+            it->value = it->in_container_index;
+            return (it->has_value = true);
+        }
+        case ARRAY_CONTAINER_TYPE: {
+            if (--it->in_container_index < 0) {
+                return (it->has_value = false);
+            }
+            const array_container_t *ac = const_CAST_array(c);
+            it->value = ac->array[it->in_container_index];
+            return (it->has_value = true);
+        }
+        case RUN_CONTAINER_TYPE: {
+            if (it->value == 0) {
+                return (it->has_value = false);
+            }
+
+            const run_container_t *rc = const_CAST_run(c);
+            it->value--;
+            if (it->value >= rc->runs[it->run_index].value) {
+                return (it->has_value = true);
+            }
+
+            if (--it->run_index < 0) {
+                return (it->has_value = false);
+            }
```

Comment:
It's probably already the case, but we should document that it's illegal to call `roaring_advance_uint32_iterator`/`roaring_previous_uint32_iterator` on an iterator with `has_value == false`.

e.g. if `container_iterator_prev` is called on a run container with one run from [5, 10], currently at 5:

1. value is not 0
2. value will be decremented to 4
3. value will be less than the current run's start
4. run_index will be decremented to -1
5. has_value will be set to false

If prev is called again:

1. value is not 0
2. value will be decremented to 3
3. we will access the run at `runs[run_index]`, which is `runs[-1]`, which is OOB

Reply:
Added documentation.